File: JANG_SO_HYEON/Programmers/SkillTree.py
# 나의 코드
def solution(skill, skill_trees):
    # 선행 스킬 : 어떤 스킬을 배우기 전에 먼저 배워야 하는 스킬
    answer = 0
    for tree in skill_trees:
        temp = 0
        for s in tree:
            if s in skill:
                if skill.index(s) == temp:
                    temp += 1
                else:
                    break
        # 중간에 break 등으로 끊기지 않을 경우 else문이 실행됨
        else:
            answer += 1

    return answer

File: PARK_YOUNG_CHEOL/src/programmers/summerWinterCoding2018/SkillTree.java
package programmers.summerWinterCoding2018;

import java.util.LinkedList;
import java.util.Queue;

/**
 * 13분
 */
public class SkillTree {

    static int solution(String skill, String[] skill_trees) {
        int answer = 0;

        Queue<Character> userQ = new LinkedList<Character>();

        for (String s : skill_trees) {
            for (int i = 0; i < s.length(); i++) {
                if (skill.contains(s.charAt(i)+"")){
                    userQ.offer(s.charAt(i));
                }
            }

            // 큐와 비교함
            boolean isCountable = true;
            int pos = 0;
            while (!userQ.isEmpty()){
                if (pos < skill.length() && userQ.poll()==skill.charAt(pos)){
                    pos++;
                } else {
                    isCountable = false;
                    break;
                }
            }

            if (isCountable) answer++;
            userQ.clear();
        }

        return answer;
    }

    public static void main(String[] args) {
        String skill = "CBD";
        String[] skill_trees = {"BACDE", "CBADF", "AECB", "BDA"};

        System.out.println(solution(skill, skill_trees));
    }
}
